pkg/kubelet/pleg/pleg.go

```diff
@@ -63,7 +64,7 @@ type PodLifecycleEvent struct {
 
 // PodLifecycleEventGenerator contains functions for generating pod life cycle events.
 type PodLifecycleEventGenerator interface {
-	Start()
+	Start(context.Context)
```

Comment:
Why did that not cause problems when it was missing? Is this interface actually used for anything?

Reply:
Before being fixed, it caused an error here when `EventedPLEG` is enabled:
https://github.com/kubernetes/kubernetes/blob/bbd83d86444d7b325a51f4daa0d65163b795b70e/pkg/kubelet/pleg/evented.go#L92-L95

Reply:
That's some odd piece of code. If NewEventedPLEG needs a podLifecycleEventGeneratorHandler, why is that not part of the official API?

Reply:
I don't think any of those interfaces are needed. They can be removed and internal methods can be made un-exported, which further helps to understand what is part of the API and what isn't. Here's a patch against current master:

```patch
diff --git a/pkg/kubelet/kubelet.go b/pkg/kubelet/kubelet.go
index 1791a663a10..720e5fff145 100644
--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -1217,10 +1217,10 @@ type Kubelet struct {
 	// pleg observes the state of the container runtime and notifies the kubelet of changes to containers, which
 	// notifies the podWorkers to reconcile the state of the pod (for instance, if a container dies and needs to
 	// be restarted).
-	pleg pleg.PodLifecycleEventGenerator
+	pleg *pleg.GenericPLEG
 
 	// eventedPleg supplements the pleg to deliver edge-driven container changes with low-latency.
-	eventedPleg pleg.PodLifecycleEventGenerator
+	eventedPleg *pleg.EventedPLEG
 
 	// Store kubecontainer.PodStatus for all pods.
 	podCache kubecontainer.Cache
diff --git a/pkg/kubelet/pleg/evented.go b/pkg/kubelet/pleg/evented.go
index 2a1b70e3f0c..c18435198ae 100644
--- a/pkg/kubelet/pleg/evented.go
+++ b/pkg/kubelet/pleg/evented.go
@@ -71,8 +71,8 @@ type EventedPLEG struct {
 	cache kubecontainer.Cache
 	// For testability.
 	clock clock.Clock
-	// GenericPLEG is used to force relist when required.
-	genericPleg podLifecycleEventGeneratorHandler
+	// genericPLEG is used to force relist when required.
+	genericPleg *GenericPLEG
 	// The maximum number of retries when getting container events from the runtime.
 	eventedPlegMaxStreamRetries int
 	// Indicates relisting related parameters
@@ -87,18 +87,14 @@ type EventedPLEG struct {
 
 // NewEventedPLEG instantiates a new EventedPLEG object and return it.
 func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
-	cache kubecontainer.Cache, genericPleg PodLifecycleEventGenerator, eventedPlegMaxStreamRetries int,
-	relistDuration *RelistDuration, clock clock.Clock) (PodLifecycleEventGenerator, error) {
-	handler, ok := genericPleg.(podLifecycleEventGeneratorHandler)
-	if !ok {
-		return nil, fmt.Errorf("%v doesn't implement podLifecycleEventGeneratorHandler interface", genericPleg)
-	}
+	cache kubecontainer.Cache, genericPleg *GenericPLEG, eventedPlegMaxStreamRetries int,
+	relistDuration *RelistDuration, clock clock.Clock) (*EventedPLEG, error) {
 	return &EventedPLEG{
 		runtime:                     runtime,
 		runtimeService:              runtimeService,
 		eventChannel:                eventChannel,
 		cache:                       cache,
-		genericPleg:                 handler,
+		genericPleg:                 genericPleg,
 		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
 		relistDuration:              relistDuration,
 		clock:                       clock,
@@ -110,8 +106,8 @@ func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
 	return e.eventChannel
 }
 
-// Relist relists all containers using GenericPLEG
-func (e *EventedPLEG) Relist() {
+// relist relists all containers using GenericPLEG
+func (e *EventedPLEG) relist() {
 	e.genericPleg.Relist()
 }
 
@@ -129,8 +125,8 @@ func (e *EventedPLEG) Start() {
 	go wait.Until(e.updateGlobalCache, globalCacheUpdatePeriod, e.stopCacheUpdateCh)
 }
 
-// Stop stops the Evented PLEG
-func (e *EventedPLEG) Stop() {
+// stop stops the Evented PLEG
+func (e *EventedPLEG) stop() {
 	e.runningMu.Lock()
 	defer e.runningMu.Unlock()
 	if !isEventedPLEGInUse() {
@@ -150,9 +146,9 @@ func (e *EventedPLEG) updateGlobalCache() {
 	e.cache.UpdateTime(time.Now())
 }
 
-// Update the relisting period and threshold
-func (e *EventedPLEG) Update(relistDuration *RelistDuration) {
-	e.genericPleg.Update(relistDuration)
+// update the relisting period and threshold
+func (e *EventedPLEG) update(relistDuration *RelistDuration) {
+	e.genericPleg.update(relistDuration)
 }
 
 // Healthy check if PLEG work properly.
@@ -185,9 +181,9 @@ func (e *EventedPLEG) watchEventsChannel() {
 				if isEventedPLEGInUse() {
 					// Fall back to Generic PLEG relisting since Evented PLEG is not working.
 					klog.V(4).InfoS("Fall back to Generic PLEG relisting since Evented PLEG is not working")
-					e.Stop()
-					e.genericPleg.Stop()       // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
-					e.Update(e.relistDuration) // Update the relisting period to the default value for the Generic PLEG.
+					e.stop()
+					e.genericPleg.stop()       // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
+					e.update(e.relistDuration) // Update the relisting period to the default value for the Generic PLEG.
 					e.genericPleg.Start()
 					break
 				}
@@ -199,7 +195,7 @@ func (e *EventedPLEG) watchEventsChannel() {
 			if err != nil {
 				metrics.EventedPLEGConnErr.Inc()
 				numAttempts++
-				e.Relist() // Force a relist to get the latest container and pods running metric.
+				e.relist() // Force a relist to get the latest container and pods running metric.
 				klog.V(4).InfoS("Evented PLEG: Failed to get container events, retrying: ", "err", err)
 			}
 		}
diff --git a/pkg/kubelet/pleg/generic.go b/pkg/kubelet/pleg/generic.go
index a020101920c..c8997bdc7dd 100644
--- a/pkg/kubelet/pleg/generic.go
+++ b/pkg/kubelet/pleg/generic.go
@@ -118,7 +118,7 @@ type podRecords map[types.UID]*podRecord
 // NewGenericPLEG instantiates a new GenericPLEG object and return it.
 func NewGenericPLEG(runtime kubecontainer.Runtime, eventChannel chan *PodLifecycleEvent,
 	relistDuration *RelistDuration, cache kubecontainer.Cache,
-	clock clock.Clock) PodLifecycleEventGenerator {
+	clock clock.Clock) *GenericPLEG {
 	return &GenericPLEG{
 		relistDuration: relistDuration,
 		runtime:        runtime,
@@ -147,7 +147,7 @@ func (g *GenericPLEG) Start() {
 	}
 }
 
-func (g *GenericPLEG) Stop() {
+func (g *GenericPLEG) stop() {
 	g.runningMu.Lock()
 	defer g.runningMu.Unlock()
 	if g.isRunning {
@@ -156,7 +156,7 @@ func (g *GenericPLEG) Stop() {
 	}
 }
 
-func (g *GenericPLEG) Update(relistDuration *RelistDuration) {
+func (g *GenericPLEG) update(relistDuration *RelistDuration) {
 	g.relistDuration = relistDuration
 }
 
diff --git a/pkg/kubelet/pleg/pleg.go b/pkg/kubelet/pleg/pleg.go
index 0a44745925b..9f80b99d884 100644
--- a/pkg/kubelet/pleg/pleg.go
+++ b/pkg/kubelet/pleg/pleg.go
@@ -20,7 +20,6 @@ import (
 	"time"
 
 	"k8s.io/apimachinery/pkg/types"
-	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
 // PodLifeCycleEventType define the event type of pod life cycle events.
@@ -60,20 +59,3 @@ type PodLifecycleEvent struct {
 	//   - All other event types: unused.
 	Data interface{}
 }
-
-// PodLifecycleEventGenerator contains functions for generating pod life cycle events.
-type PodLifecycleEventGenerator interface {
-	Start()
-	Watch() chan *PodLifecycleEvent
-	Healthy() (bool, error)
-	UpdateCache(*kubecontainer.Pod, types.UID) (error, bool)
-}
-
-// podLifecycleEventGeneratorHandler contains functions that are useful for different PLEGs
-// and need not be exposed to rest of the kubelet
-type podLifecycleEventGeneratorHandler interface {
-	PodLifecycleEventGenerator
-	Stop()
-	Update(relistDuration *RelistDuration)
-	Relist()
-}
```

Reply:
There was some discussion around these interfaces in the https://github.com/kubernetes/kubernetes/pull/113825. Not sure its related, but worth reading in my opinion. [This PR](https://github.com/kubernetes/kubernetes/pull/111384) and [his comment](https://github.com/kubernetes/kubernetes/pull/111384#discussion_r1015933735) in particular may add some context to this.

Reply:
I browsed through that. I saw comments about which methods should be in these interfaces, but no explanation why they are there in the first place.

https://github.com/kubernetes/kubernetes/pull/111384#discussion_r1015933735
> as this is all the rest of the kubelet needs to know about

A simpler way to achieve this is to only export methods which really need to be called by kubelet.

Reply:
I agree. However, it should probably be in a separate PR.